src/Symfony/Component/Serializer/NameConverter/MetadataAwareNameConverter.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\NameConverter;
+
+use Symfony\Component\Serializer\Mapping\Factory\ClassMetadataFactoryInterface;
+
+/**
+ * @author Fabien Bourigault <bourigaultfabien@gmail.com>
+ */
+final class MetadataAwareNameConverter implements AdvancedNameConverterInterface
+{
+    /**
+     * @var ClassMetadataFactoryInterface
+     */
+    private $metadataFactory;
+
+    public function __construct(ClassMetadataFactoryInterface $metadataFactory)
+    {
+        $this->metadataFactory = $metadataFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($propertyName, string $class = null, string $format = null, array $context = array())
+    {
+        if (null === $class || !$this->metadataFactory->hasMetadataFor($class)) {
```

Comment:
I think the result of this check should be cached in private static property or will soon become a performance bottleneck: it will be called for every property of every objects in the graph.

Same for denormalize.

Reply:
For caching I was thinking about having a `CacheNameConverter` decorator which implements the `AdvancedNameConverterInterface` to allow caching result of the `ChainNameConverter` I want to add for the framework integration.

But isn't the `$context` array preventing me to build an efficient cache key because some decorated name converters can return different result for the same `$propertyName`, `$class` and `$format` depending on the `$context` ?

Also, any user defined name converter may be non-deterministic.

So I guess I can't have a generic `CacheNameConverter`.

But what about using a PSR-6 optional cache in the `MetadataAwareNameConverter` to have result cached between requests ? The cache key would be built using the `$propertyName` and `$cache` properties.

Reply:
After talking with @dunglas on Slack, this PR will only introduce a local static cache which improve the performances when more than one object of a class is involved in a serialization/deserialization (the common case is a "to-many" collection).

Using PSR-6 and a CacheWarmer will be done in a separate PR if required for even better performance.